internal/core/pipeline_config.go

```diff
@@ -18,10 +20,43 @@ func (p *Pipeline) ConfigSync(ctx context.Context) error {
 		return err
 	}
 
-	// TODO(briancain): do we need a Multi upsert?
 	p.logger.Debug("syncing pipeline config")
 	for _, pipeline := range pipelines {
-		_, err := p.client.UpsertPipeline(ctx, &pb.UpsertPipelineRequest{
+		pipeOwner, ok := pipeline.Owner.(*pb.Pipeline_Project)
+		if !ok {
+			return status.Error(codes.FailedPrecondition,
+				"failed to determine pipeline owner. Only expecting pipeline owner to be a Pipeline_Project type.")
+		}
+
+		p.logger.Trace("looking up pipeline by owner and name", "owner", pipeOwner.Project, "name", pipeline.Name)
+
+		// Look up if this pipeline already exists by Pipeline Owner (project and pipeline name)
+		// If so, use the existing id to upsert into an existing config. Otherwise
+		// create a new pipeline entry.
```

Comment:
I think we should do this automatically on the backend. We can chat about it on Monday but that way I think we can do this all in a single transaction.

Reply:
Roger that. That shouldn't be too difficult to add here!

Reply:
I think that's covered via [4789fa5](https://github.com/hashicorp/waypoint/pull/3397/commits/4789fa5b7338b2eef6c3682a1e34091ab081b174) now